src/common/proto_type.cpp

```diff
@@ -46,32 +48,46 @@ ProtoType ProtoType::of_value(Value value) {
     }
 };
 
-Struct map_to_struct(std::unordered_map<std::string, ProtoType*> dict) {
+Struct map_to_struct(std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>> dict) {
     Struct s;
-    for (auto& key_and_value : dict) {
+    for (auto& key_and_value : *dict) {
         const std::string key = key_and_value.first;
         Value value = key_and_value.second->proto_value();
+        const std::string stg = boost::get<std::string>(key_and_value.second->proto_type);
+
         google::protobuf::MapPair<std::string, Value> val(key, value);
+            
         s.mutable_fields()->insert(val);
     }
 
     return s;
 }
 
-std::unordered_map<std::string, ProtoType*> struct_to_map(Struct struct_) {
+
+
+std::shared_ptr<std::unordered_map<std::string, std::shared_ptr<ProtoType>>> struct_to_map(Struct struct_) {
     google::protobuf::Map<std::string, Value> struct_map = struct_.fields();
-    std::unordered_map<std::string, ProtoType*> map;
+    std::unordered_map<std::string, std::shared_ptr<ProtoType>> map;
+
     for (auto& val : struct_.fields()) {
         std::string key = val.first;
-        ProtoType value = ProtoType::of_value(val.second);
-        map.emplace(key, &value);
+        std::shared_ptr<Value> val_ptr = std::make_shared<Value>(val.second); 
+        std::shared_ptr<ProtoType> value = std::make_shared<ProtoType>(ProtoType::of_value(val.second));
+        map.emplace(key, value);
     }
 
-    return map;
+    return std::make_shared<std::unordered_map<std::string, std::shared_ptr<ProtoType>>>(map);
+    
 };
 
 Value ProtoType::proto_value() {
     Value v;
+       try {
```

Comment:
We already have a default that throws with the message `Invalid proto_value conversion type. This should never happen; please file a bug report.";`. Is it good if I remove the try catch and keep that error or should I change it to runtime_exception?

Reply:
Oh, you're absolutely right, we do already have the default case! 🤦 thanks for catching that :)

In that case, I think we should get rid of this `try catch` here and change the `throw` at the bottom to a `runtime_exception` instead of a string.